File: Spielfeld.ts
import { Gegenstand } from "./Gegenstand"
import { Monster } from "./Monster"

export enum Typ {
    Gang,
    Ecke,
    TStück,
    Kreuzung
}

export enum Orientierung {
    Norden,
    Osten,
    Süden,
    Westen
}

export enum Ausstattung {
    Heilquelle,
    Portal,
    Raum
}

export class Raumkarte {
    typ: Typ
    ausstattung: (false | Ausstattung)

    constructor(typ: Typ, ausstattung: (false | Ausstattung)) {
        this.typ = typ
        this.ausstattung = ausstattung
    }
}

type Raum = {
    norden: boolean,
    osten: boolean,
    süden: boolean,
    westen: boolean,
    inhalt: (false | Ausstattung | Monster | Gegenstand )
}
export class Spielfeld {
    feld: { [position: string]: Raum } = {}


    public setFeld(x: number, y: number, raum:  Raum) {
        this.feld[`${x}|${y}`] = raum
    }

    public getFeld(x: number, y: number): Raum {
        return this.feld[`${x}|${y}`]
    }

    lege_karte(x: number, y: number, raumkarte: Raumkarte, orientierung: Orientierung) {
        const raum = {
            norden: true,
            osten: true,
            süden: true,
            westen: true,
            inhalt: raumkarte.ausstattung
        }
        switch (raumkarte.typ) {
            case Typ.Gang: {
                if ([Orientierung.Norden, Orientierung.Süden].includes(orientierung)) {
                    raum.osten = false
                    raum.westen = false
                } else {
                    raum.norden = false
                    raum.süden = false
                }
                break
            }
            case Typ.Ecke: {
                switch (orientierung) {
                    case Orientierung.Norden: {
                        raum.süden = false
                        raum.westen = false
                        break
                    }
                    case Orientierung.Osten: {
                        raum.westen = false
                        raum.norden = false
                        break
                    }
                    case Orientierung.Süden: {
                        raum.norden = false
                        raum.osten = false
                        break
                    }
                    case Orientierung.Westen: {
                        raum.osten = false
                        raum.süden = false
                        break
                    }
                }
                break
            }
            case Typ.TStück: {
                switch (orientierung) {
                    case Orientierung.Norden: {
                        raum.westen = false
                        break
                    }
                    case Orientierung.Osten: {
                        raum.norden = false
                        break
                    }
                    case Orientierung.Süden: {
                        raum.osten = false
                        break
                    }
                    case Orientierung.Westen: {
                        raum.süden = false
                        break
                    }
                }
                break
            }
        }
        this.setFeld(x, y, raum)
    }

    raum_ziehe_inhalt(x: number, y: number, monster: Monster) {
        this.getFeld(x, y).inhalt = monster
    }
    monster_besiegt(x: number, y: number) {
        this.getFeld(x, y).inhalt = (this.getFeld(x, y).inhalt as Monster).loot
    }

    gegenstand_aufgenommen(x: number, y: number) {
        this.getFeld(x, y).inhalt = false
    }

    gegenstand_abgelegt(x: number, y: number, gegenstand: Gegenstand) {
        this.getFeld(x, y).inhalt = gegenstand
    }
}

File: Spielfeld.py
import typing
from enum import Enum
from Gegenstand import Gegenstand
from Monster import Monster

Position = typing.Tuple[int, int]


class Typ(Enum):
    Gang = 1
    Ecke = 2
    TStück = 3
    Kreuzung = 4

class Orientierung(Enum):
    Norden = 1
    Osten = 2
    Süden = 3
    Westen = 4

class Ausstattung(Enum):
    Heilquelle = 1
    Portal = 2
    Raum = 3

class Raumkarte:
    typ: Typ
    ausstattung: typing.Optional[Ausstattung]

    def __init__(self, typ: Typ, ausstattung: typing.Optional[Ausstattung]):
        self.typ = typ
        self.ausstattung = ausstattung

class Raum(typing.TypedDict):
    norden: bool
    osten: bool
    süden: bool
    westen: bool
    inhalt: typing.Optional[typing.Union[Ausstattung, Monster, Gegenstand]]

class Spielfeld:
    feld: typing.Dict[Position, Raum] = {}

    def __getitem__(self, key: Position) -> typing.Union[Raum, None]:
        return self.feld.get(key, None)

    def __setitem__(self, key: Position, raum: Raum):
        self.feld[key] = raum

    def lege_karte(self, pos: Position, raumkarte: Raumkarte, orientierung: Orientierung) -> None:
        raum: Raum = {
            'norden': True,
            'osten': True,
            'süden': True,
            'westen': True,
            'inhalt': raumkarte.ausstattung}

        if raumkarte.typ == Typ.Gang:
            if orientierung in [Orientierung.Norden, Orientierung.Süden]:
                raum['osten'] = False
                raum['westen'] = False
            else:
                raum['norden'] = False
                raum['süden'] = False
        elif raumkarte.typ == Typ.Ecke:
            if orientierung == Orientierung.Norden:
                raum['süden'] = False
                raum['westen'] = False
            elif orientierung == Orientierung.Osten:
                raum['westen'] = False
                raum['norden'] = False
            elif orientierung == Orientierung.Süden:
                raum['norden'] = False
                raum['osten'] = False
            else:
                raum['osten'] = False
                raum['süden'] = False
        elif raumkarte.typ == Typ.TStück:
            if orientierung == Orientierung.Norden:
                raum['westen'] = False
            elif orientierung == Orientierung.Osten:
                raum['norden'] = False
            elif orientierung == Orientierung.Süden:
                raum['osten'] = False
            else:
                raum['süden'] = False
        self.feld[pos] = raum

    def raum_ziehe_inhalt(self, pos: Position, monster: Monster) -> None:
        self.feld[pos]['inhalt'] = monster

    def monster_besiegt(self, pos: Position) -> None:
        self.feld[pos]['inhalt'] = typing.cast(Monster, self.feld[pos]['inhalt']).loot

    def gegenstand_aufgenommen(self, pos: Position) -> None:
        self.feld[pos]['inhalt'] = None

    def gegenstand_abgelegt(self, pos: Position, gegenstand: Gegenstand):
        self.feld[pos]['inhalt'] = gegenstand
